Pause on exception

andrewfulton9 opened the pull request:
<!--
Thanks for contributing to JupyterLab!
Please fill out the following items to submit a pull request.
See the contributing guidelines for more information:
https://github.com/jupyterlab/jupyterlab/blob/master/CONTRIBUTING.md
-->

## References

This is a continuation of PR draft #10298.
It references issue #9544

<!-- Note issue numbers this pull request addresses (should be at least one, see contributing guidelines above). -->

<!-- Note any other pull requests that address this issue and how this pull request is different. -->

## Code changes

<!-- Describe the code changes and how they address the issue. -->

## User-facing changes
![chrome-capture (3)](https://user-images.githubusercontent.com/13774419/147505579-5758333d-ff5a-4ce7-b24d-4e1bc8276c4f.gif)

<!-- Describe any visual or user interaction changes and how they address the issue. -->

<!-- For visual changes, include before and after screenshots here. -->

## Backwards-incompatible changes

<!-- Describe any backwards-incompatible changes to JupyterLab public APIs. -->

fcollonval commented on packages/debugger-extension/src/index.ts:
```diff
@@ -695,13 +696,34 @@ const main: JupyterFrontEndPlugin<void> = {
     });
 
+    commands.addCommand(CommandIDs.pause, {
+      label: trans.__('Enable / Disable pausing on exceptions'),
```
You could improve it by using the status:
```suggestion
      label: service.isPausingOnExceptions ? trans.__('Disable pausing on exceptions') : trans.__('Enable pausing on exceptions'),
```

fcollonval commented on packages/debugger/src/panels/callstack/index.ts:
```diff
@@ -81,6 +81,15 @@ export class Callstack extends PanelWithToolbar {
       })
     );
 
+    this.toolbar.addItem(
```
I would prefer the button to appear on the breakpoints panel not on the callstack one.

fcollonval commented on packages/debugger/src/panels/callstack/index.ts:
```diff
@@ -128,6 +137,11 @@ export namespace Callstack {
     evaluate: string;
+
+    /**
+     * The puse command ID.
```
```suggestion
     * The pause command ID.
```

fcollonval commented on packages/debugger-extension/src/index.ts:
```diff
@@ -695,13 +696,34 @@ const main: JupyterFrontEndPlugin<void> = {
+        return service.isPausingOnExceptions;
+      },
+      isEnabled: () => {
+        return !!service.isStarted;
```
What about kernel not supporting that type of exceptions?

andrewfulton9 replied:
At worst, I think the kernel would just run until it hit the the exception instead of breaking immediately before. That being said, pauseOnExceptions sets the breakpoint argument options via the debug protocol so my assumption is that it should be kernel agnostic assuming their is a debug protocol for it. I have only tried running it with the ipython and xeus python kernels though.

fcollonval replied:
I understand but the debug protocol does specify an exchange of capabilities at start-up. Example for ipykernel (server reply):

![image](https://user-images.githubusercontent.com/8435071/148030507-e12a1fdc-75e7-43a9-8f75-e5e58fbe250c.png)

So it seems, we do have a way to determine if the kernel supports it.

andrewfulton9 replied:
Those are currently used to build the options object that gets sent to debugger. So if none are defined for a kernel then pause on exception would never pause. Otherwise it should stop on those breakpoint filters.

fcollonval replied:
From the user point of view, this means the button will not do anything what should be avoided. This is why it will be nice if the `isEnabled` or perhaps the `isVisible` attribute of the command will be set accordingly.

fcollonval commented on packages/debugger/src/service.ts:
```diff
@@ -498,6 +513,34 @@ export class DebuggerService implements IDebugger, IDisposable {
+    const filters: string[] = [];
+    const exceptionOptions: DebugProtocol.ExceptionOptions[] = [];
+    const breakMode = enable ? 'userUnhandled' : 'never';
+    for (let filter_dict of exceptionBreakpointFilters ?? []) {
```
`filter_dict` is not a valid identifier. You should use camel case.

fcollonval commented on packages/debugger/src/session.ts:
```diff
@@ -80,12 +82,42 @@ export class DebuggerSession implements IDebugger.ISession {
+  get pausingOnExceptions(): boolean {
+    return this._pausingOnExceptions;
+  }
+
+  /**
+   * Sets pause on exception
+   */
```
When you define a getter and a setter only one doc string is required.

```suggestion
```

fcollonval commented on packages/debugger/src/tokens.ts:
```diff
@@ -46,11 +51,19 @@ export interface IDebugger {
   readonly sessionChanged: ISignal<IDebugger, IDebugger.ISession | null>;
 
+  /**
+   * TODO: Signal emitted upon session changed.
```
`TODO` to resolve

fcollonval commented on packages/debugger/src/tokens.ts:
```diff
@@ -46,11 +51,19 @@ export interface IDebugger {
    */
   clearBreakpoints(): Promise<void>;
 
+  /** TODO */
```
`TODO` to resolve

fcollonval commented on packages/debugger/style/icons.css:
```diff
@@ -25,3 +25,13 @@ button.jp-Button.jp-mod-minimal.jp-TableView:hover {
 }
+
+/* Pause on exceptions */
+button.jp-Button.bp3-button.bp3-minimal.jp-PauseOnExceptions:hover {
```
Blueprint has been removed

```suggestion
button.jp-Button.jp-mod-minimal.jp-PauseOnExceptions:hover {
```

fcollonval commented on packages/debugger-extension/src/index.ts:
```diff
@@ -695,13 +696,34 @@ const main: JupyterFrontEndPlugin<void> = {
       }
     });
 
+    service.stateRestored.connect(_ => {
```
(Question due to my lack of knowledge)

Is this needed? Is the connector below not doing the job already?

andrewfulton9 replied:
You are right. I'll get that taken out.

andrewfulton9 commented:
@fcollonval, I think I have addressed everything from your previous review, and fixed an issue with the pause button not updating when switching tabs. Could you give it another review when you get a chance? Thank you!

fcollonval reviewed (changes requested):
This looks great @andrewfulton9 (tested on Binder)

I have two minor comments and it should be good to go.

andrewfulton9 commented:
@fcollonval, I added the docstring. Let me know if there is anything else I should do. I think this should be good to go otherwise. I rebased on master. I think thats where all the auto tags are coming from. I'm not sure if they are supposed to be added.

fcollonval commented:
@andrewfulton9 would you mind rebasing this PR? It looks like something went wrong hence the amount of changes (and all those tags).

andrewfulton9 commented:
@fcollonval, looks like re-basing again did the trick. I'm not sure what happened on the first one. Thanks for the suggestion!

fcollonval commented on packages/debugger/package.json:
```diff
@@ -73,7 +73,8 @@
     "@vscode/debugprotocol": "^1.51.0",
     "codemirror": "~5.61.0",
-    "react": "^17.0.1"
+    "react": "^17.0.1",
+    "vscode-debugprotocol": "~1.51.0"
```
This probably slips in following the rebase. `vscode-debugprotocol` has been renamed `@vscode/debugprotocol` and [that change](https://github.com/jupyterlab/jupyterlab/pull/11812) is already on master.

```suggestion
    "react": "^17.0.1"
```

fcollonval commented on packages/debugger/src/session.ts:
```diff
@@ -9,6 +9,8 @@ import { PromiseDelegate } from '@lumino/coreutils';
 
 import { ISignal, Signal } from '@lumino/signaling';
 
+import { DebugProtocol } from 'vscode-debugprotocol';
```
```suggestion
import { DebugProtocol } from '@vscode/debugprotocol';
```

fcollonval commented on packages/debugger/src/service.ts:
```diff
@@ -859,6 +950,7 @@ export class DebuggerService implements IDebugger, IDisposable {
   private _sessionChanged = new Signal<IDebugger, IDebugger.ISession | null>(
     this
   );
+  private _stateRestored = new Signal<IDebugger, null>(this);
```
It looks like `stateRestored` signal is not read in the final version. So let's get ride of it.

```suggestion
```

fcollonval reviewed (changes requested):
@andrewfulton9 sorry two more changes are required. Following the rebase `vscode-debugprotocol` package is now named `@vscode/debugprotocol`. And it looks like we can remove `stateRestored` signal as it is not read anywhere.

andrewfulton9 commented:
@fcollonval, thanks for catching those! I committed all your suggestions.

fcollonval commented:
@andrewfulton9 the *Linux Tests / Linux (release_test, 3.10) (pull_request)* is legit. The yarn.lock needs to be updated (most likely to remove the old `vscode-debugprotocol` package). Could you please update the PR?

andrewfulton9 commented:
@fcollonval, thanks for pointing me in the right direction. I got that fixed. That last failed isn't something to be fixed in this PR, right?

fcollonval reviewed (approved):
Thanks @andrewfulton9 for pushing this through

jtpio commented:
Thanks!

CI is now all green.

jtpio commented:
@meeseeksdev please backport to 3.3.x

andrewfulton9 commented:
@jtpio, Yeah I'll try to do that today

krassowski commented:
I updated the top-level comment with recent screenshots (as the button was moved to a different location). Was there a discussion about having the caption differ from label?

```ts
       label: service.isPausingOnExceptions
        ? trans.__('Disable pausing on exceptions')
        : trans.__('Enable pausing on exceptions'),
      caption: trans.__('Enable / Disable pausing on exceptions'),
```

It would be easier for me to know if the option is on/off if it was separated.

fcollonval commented:
> It would be easier for me to know if the option is on/off if it was separated.

Good point I missed it when reviewing. I opened #11970
